runtime/compiler/runtime/JITServerAOTCache.cpp

```diff
@@ -1429,17 +1532,276 @@ JITServerAOTCacheMap::~JITServerAOTCacheMap()
+                  }
+               }
+            else // Saving the AOT cache to file failed
+               {
+               fclose(newCacheFile);
+               newCacheFile = NULL;
+               if (TR::Options::getVerboseOption(TR_VerboseJITServer))
+                  TR_VerboseLog::writeLineLocked(TR_Vlog_JITServer, "AOT cache: Failed to serialize cache '%s' to file", cacheName.c_str());
+               remove(tempFileName.c_str());
+               }
+            }
+         else  // Cannot open file for writing; permissions?
+            {
+            if (TR::Options::getVerboseOption(TR_VerboseJITServer))
+               TR_VerboseLog::writeLineLocked(TR_Vlog_JITServer, "AOT cache: Cannot open new file %s for writing", tempFileName.c_str());
+            cache->excludeCacheFromSavingToFile();
+            }
+         }
+      } // end try
+   catch(const std::exception& e)
+      {
+      if (TR::Options::getVerboseOption(TR_VerboseJITServer))
+         {
+         TR_VerboseLog::writeLineLocked(TR_Vlog_JITServer, "AOT cache: exception caught when trying to save cache '%s': %s", cacheName.c_str(), e.what());
+         }
+      }
+
+   cache->finalizeSaveOperation(success, numAOTMethodsWritten);
+   return success;
+   }
+
+
+void
+JITServerAOTCacheMap::loadNextQueuedAOTCacheFromFile(J9::J9SegmentProvider &scratchSegmentProvider)
+   {
+   std::string cacheName;
+      {
+      OMR::CriticalSection cs(_monitor);
+      TR_ASSERT(_cachesBeingLoaded.size() != 0, "There must be at least one AOT cache waiting to be loaded from file");
+
+      // Extract the first entry from the queue
+      // If there is nothing in the queue, then just return (another thread is working on it)
+      if (_cachesToLoadQueue.empty())
+         return;
+      cacheName = _cachesToLoadQueue.front();
+      _cachesToLoadQueue.pop_front();
+      }
+   TR::CompilationInfo *compInfo = TR::CompilationInfo::get();
+   std::string cacheFileName = buildCacheFileName(compInfo->getPersistentInfo()->getJITServerAOTCacheDir(), cacheName);
+   JITServerAOTCache *cache = NULL;
+
+   // Open the AOT cache file and create a new JITServerAOTCache object
+   FILE *cacheFile = fopen(cacheFileName.c_str(), "rb");
+   if (cacheFile)
+      {
+      if (TR::Options::getVerboseOption(TR_VerboseJITServer))
+         TR_VerboseLog::writeLineLocked(TR_Vlog_JITServer, "AOT cache: t=%llu Opened file %s to load cache '%s' from file",
+                                        compInfo->getPersistentInfo()->getElapsedTime(), cacheFileName.c_str(), cacheName.c_str());
+      size_t segmentSize = scratchSegmentProvider.getPreferredSegmentSize();
+      if (!segmentSize)
+         segmentSize = 1 << 24/*16 MB*/;
+      TR::RawAllocator rawAllocator(compInfo->getJITConfig()->javaVM);
+      J9::SystemSegmentProvider segmentProvider(1 << 16/*64 KB*/, segmentSize, TR::Options::getScratchSpaceLimit(), scratchSegmentProvider, rawAllocator);
```

Comment:
It is worth noting that the constructor of `J9::SystemSegmentProvider` can throw in which case we won't close the `cacheFile` handle.

Reply:
I extended the try/catch block to cover that area of code as well